src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CreateComponentCodeActionResolver.cs

```diff
@@ -0,0 +1,111 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Extensions;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor;
+using Newtonsoft.Json.Linq;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions
+{
+    internal class CreateComponentCodeActionResolver : RazorCodeActionResolver
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly DocumentResolver _documentResolver;
+
+        public CreateComponentCodeActionResolver(
+            ForegroundDispatcher foregroundDispatcher,
+            DocumentResolver documentResolver)
+        {
+            _foregroundDispatcher = foregroundDispatcher ?? throw new ArgumentNullException(nameof(foregroundDispatcher));
+            _documentResolver = documentResolver ?? throw new ArgumentNullException(nameof(documentResolver));
+        }
+
+        public override string Action => LanguageServerConstants.CodeActions.CreateComponentFromTag;
+
+        public override async Task<WorkspaceEdit> ResolveAsync(JObject data, CancellationToken cancellationToken)
+        {
+            var actionParams = data.ToObject<CreateComponentParams>();
+            var path = actionParams.Uri.GetAbsoluteOrUNCPath();
+
+            var document = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(path, out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+            if (document is null)
+            {
+                return null;
+            }
+
+            var text = await document.GetTextAsync().ConfigureAwait(false);
+            if (text is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            if (!FileKinds.IsComponent(codeDocument.GetFileKind()))
+            {
+                return null;
+            }
+
+            var newComponentUri = new UriBuilder()
+            {
+                Scheme = Uri.UriSchemeFile,
+                Path = actionParams.Path,
+                Host = string.Empty,
+            }.Uri;
+
+            var documentChanges = new List<WorkspaceEditDocumentChange>
+            {
+                new WorkspaceEditDocumentChange(new CreateFile() { Uri = newComponentUri.ToString() })
+            };
+
+            var syntaxTree = codeDocument.GetSyntaxTree();
```

Comment:
Given this entire bottom section is a bout detecting a namespace directive and adding one for the created file i'd refactor that into its own method `TryAddNamespaceDirective`